packages/blockchain-wallet-v4-frontend/src/data/components/fundRecovery/sagas.ts

```diff
@@ -0,0 +1,178 @@
+        const dust = yield call(api.getBchDust)
+        const script = dust.output_script
+        const coinDust = Coin.fromJS({ ...dust, script })
+        const tx = bchSign(network, coinDust, selection)
+
+        yield call(api.pushBchTx, tx.txHex, dust.lock_secret)
+        yield put(A.recoverFundsSuccess(coin))
+        yield put(actions.modals.closeAllModals())
+        yield put(actions.alerts.displaySuccess(`Funds recovered to ${recoveryAddress}.`))
+        yield put(actions.components.refresh.refreshClicked())
+      } else {
+        yield put(A.recoverFundsFailure(coin, `No recovery method for ${coin}`))
+      }
+    } catch (e) {
+      const error = errorHandler(e)
+      yield put(A.recoverFundsFailure(coin, error))
+      yield put(actions.modals.closeAllModals())
+      yield put(actions.alerts.displayWarning(error || `Recovery failed for ${coin}.`))
+    }
+  }
+
+  const searchChainForFunds = function* (action: ReturnType<typeof A.searchChain>) {
+    const { payload } = action
+    const { accountIndex, coin, derivationType } = payload
+    try {
+      yield put(A.searchChainLoading(accountIndex, coin, derivationType))
+
+      if (coin === 'BCH') {
+        // FOR CHANGE
+        // 1. get the change_index from the good xpub on bch chain
+        // 2. derive change addresses from '1/change_index...0' from SEGWIT xpub
+        // 3. get unspents from the change addresses
+        // FOR WRONG RECEIVE
+        // 1. get the SEGWIT derivation
+        // 2. get unspents from SEGWIT xpub
+        // Note: wrong receive was always index 0 because of a separate bug
+        // which caused wallet to look up receive_index from the wrong xpub
+        // or, elsing things to eventually return receive_index as 0.
+
+        const wallet = selectors.core.wallet.getWallet(yield select())
+        const accounts = Wallet.selectHDAccounts(wallet)
+        const account = HDAccountList.selectAccount(accountIndex, accounts)
+        const badDerivation = account.derivations.find((d) => d.type === derivationType)
+        const goodDerivation = account.derivations.find((d) => d.type === 'legacy')
+        const badXpub = badDerivation.xpub
+        const goodXpub = goodDerivation.xpub
+        const { addresses } = yield call(api.fetchBchData, [goodXpub])
+        const { change_index } = addresses.find(({ address }) => address === goodXpub)
+
+        const recoveryAddress = selectors.core.common.bch
+          .getNextAvailableReceiveAddress(network, accountIndex, yield select())
+          .getOrFail('No BCH address found to recovery to.')
+
+        const badChange: string[] = []
+        for (let i = 0; i <= change_index; i += 1) {
+          const { publicKey } = Bitcoin.bip32.fromBase58(badXpub).derivePath(`1/${i}`)
+          const { address: badAddress } = Bitcoin.payments.p2pkh({ pubkey: publicKey })
+          if (badAddress) badChange.push(badAddress)
+        }
```

Comment:
@mpfluger get all the change addresses from the RIGHT change_index down to 0 because we don't know where the change went after we submitted the build with the fix.